astropy/modeling/statistic.py

```diff
@@ -6,36 +6,163 @@
 # pylint: disable=invalid-name
 import numpy as np
 
-__all__ = ['leastsquare']
+from astropy.utils.decorators import format_doc
 
+__all__ = ["leastsquare", "leastsquare_1d", "leastsquare_2d", "leastsquare_3d"]
+
+
+def leastsquare(measured_vals, updated_model, weights, *x):
+    """
+    Least square statistic, with optional weights, in N-dimensions.
+
+    .. |ndarray| replace:: `~numpy.ndarray`
+
+    Parameters
+    ----------
+    measured_vals : |ndarray| or Sequence
+        Measured data values. Will be cast to array whose
+        shape must match the array-cast of the evaluated model.
+    updated_model : :class:`~astropy.modeling.Model` instance
+        Model with parameters set by the current iteration of the optimizer.
+        when evaluated on "x", must return array of shape "measured_vals"
+    weights : |ndarray| or None
+        Array of weights to apply to each residual.
+    *x : |ndarray|
+        Independent variables on which to evaluate the model.
+
+    Returns
+    -------
+    res : float
+        The sum of least squares.
+
+    Raises
+    ------
+    ValueError
+        If shape mismatch between model and measured
+    Exception
+        If cannot evaluate model, like because incorrect number of inputs
+        or a shape broadcasting failure.
+
+    See Also
+    --------
+    :func:`~astropy.modeling.statistic.leastsquare_1d`
+    :func:`~astropy.modeling.statistic.leastsquare_2d`
+    :func:`~astropy.modeling.statistic.leastsquare_3d`
+
+    Notes
+    -----
+    Models in :mod:`~astropy.modeling` have broadcasting rules that try to
+    match inputs with outputs with Model shapes. Numpy arrays have flexible
+    broadcasting rules, so mismatched shapes can often be made compatible. To
+    ensure data matches the model we must perform shape comparison and leverage
+    the Numpy arithmetic functions. This can obfuscate arithmetic computation
+    overrides, like with Quantities. Implement a custom statistic for more
+    direct control.
 
-def leastsquare(measured_vals, updated_model, weights, x, y=None):
     """
+
+    model_vals = updated_model(*x)
+
+    if np.shape(model_vals) != np.shape(measured_vals):
+        raise ValueError(
+            (
+                f"Shape mismatch between model ({np.shape(model_vals)})"
+                f" and measured ({np.shape(measured_vals)})"
+            )
+        )
+
+    if weights is None:
+        weights = 1.0
+
+    return np.sum(np.square(weights * np.subtract(model_vals, measured_vals)))
+
+
+# -------------------------------------------------------------------
+
+_leastsquare_XD_doc: str = """
```

Comment:
Is there any point to using a type annotation here? It's not obvious to me, but perhaps you can convince me. Otherwise I would dispense with it.